### DISCUSSION TITLE
OSError: [Errno 105] No buffer space available

### DISCUSSION THREAD

**Discussion by platinum83** [open3e/open3e#116] [Q&A] [answered]
Hallo,

gerade versuche ich opene3 auf einem raspberrypi bei mir zum Laufen zu bringen und erhalte die folgende Fehlermeldung. Ich habe bei mir eine vitocal 250a als Master und einen einen VX3 als Slave. Hat jemand diese Fehlermeldung schon einmal gesehen und kann mir sagen, was sie bedeutet?

Ich habe opene3 im virtuellen environment schon vor einigen Tagen installiert und habe nun versucht, nach den vorhabenden Komponenten zu scannen.  

bjoern@raspberrypi:~ $ source .venv/bin/activate
(.venv) bjoern@raspberrypi:~ $ sudo ip link set can0 up type can bitrate 250000
(.venv) bjoern@raspberrypi:~ $ open3e_depictSystem
read DID enums ...
3038 DIDs listed.
scan COB-IDs 0x680 to 0x6ff ...
Exception in thread Thread-49 (_main_thread_fn):
Traceback (most recent call last):
  File "/home/bjoern/.venv/lib/python3.11/site-packages/can/interfaces/socketcan/socketcan.py", line 804, in _send_once
    sent = self.socket.send(data)
           ^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 105] No buffer space available

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.11/threading.py", line 975, in run
    self._target(*self._args, **self._kwargs)
  File "/home/bjoern/.venv/lib/python3.11/site-packages/isotp/protocol.py", line 1631, in _main_thread_fn
    super().process(rx_timeout)
  File "/home/bjoern/.venv/lib/python3.11/site-packages/isotp/protocol.py", line 863, in process
    self.txfn(msg)
  File "/home/bjoern/.venv/lib/python3.11/site-packages/isotp/protocol.py", line 1691, in python_can_tx_canbus_3plus
    owner.bus.send(can.Message(arbitration_id=msg.arbitration_id, data=msg.data,
  File "/home/bjoern/.venv/lib/python3.11/site-packages/can/interfaces/socketcan/socketcan.py", line 789, in send
    sent = self._send_once(data, channel)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bjoern/.venv/lib/python3.11/site-packages/can/interfaces/socketcan/socketcan.py", line 806, in _send_once
    raise can.CanOperationError(
can.exceptions.CanOperationError: Failed to transmit: No buffer space available [Error Code 105]
Exception in thread Thread-51 (_main_thread_fn):
Traceback (most recent call last):
  File "/home/bjoern/.venv/lib/python3.11/site-packages/can/interfaces/socketcan/socketcan.py", line 804, in _send_once
    sent = self.socket.send(data)
           ^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 105] No buffer space available

Danke für eure Hilfe! Viele Grüße 

platinum

**Comment by MyHomeMyData** [comment]
@platinum83 Die Fehlermeldung deutet darauf hin, dass der CAN-Bus nicht korrekt arbeitet. Schau Dir mal diesen [Wiki-Artikel](https://github.com/open3e/open3e/wiki/020-Inbetriebnahme-CAN-Adapter-am-Raspberry) an.

Welchen Raspi-Typ verwendest Du?

**Comment by platinum83** [comment]
Hallo MyHomeMyData,

danke für den Hinweis. Ich habe einen  Raspberry Pi 4 Modell B ARM (8 GB).

Die Wiki Seite habe ich angeschaut und Vieles ausprobiert. Könntest du dir bitte den folgenden Output anschauen? vielleicht hast du noch eine Idee: 


bjoern@raspberrypi:/ $ ifconfig -a
can0: flags=128<NOARP>  mtu 16
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 10  (UNSPEC)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

bjoern@raspberrypi:/ $ dmesg | grep CAN
[    2.362095] usb 1-1.3: Product: USB2CAN V3.3
[    6.443031] CAN device driver interface
[1237396.230588] usb 1-1.3: Product: USB2CAN V3.3
[1239027.862912] NET: Registered PF_CAN protocol family
[1239926.453514] usb 1-1.3: Product: USB2CAN V3.3
[1252031.117437] usb 1-1.3: Product: USB2CAN V3.3
[1303708.081665] usb 1-1.2: Product: USB2CAN V3.3

auf der WIki Seite gab es hier nur zwei Zeilen Output. Kann es sein dass der Treiber zu oft installiert ist?


bjoern@raspberrypi:/ $ ip -details link show can0
8: can0: <NOARP,UP,LOWER_UP,ECHO> mtu 16 qdisc pfifo_fast state UP mode DEFAULT group default qlen 10
    link/can  promiscuity 0  allmulti 0 minmtu 0 maxmtu 0
    can state ERROR-ACTIVE restart-ms 0
          bitrate 250000 sample-point 0.875
          tq 250 prop-seg 6 phase-seg1 7 phase-seg2 2 sjw 1 brp 12
          gs_usb: tseg1 1..16 tseg2 1..8 sjw 1..4 brp 1..1024 brp_inc 1
          clock 48000000 numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 tso_max_size 65536 tso_max_segs 65535 gro_max_size 65536 parentbus usb parentdev 1-1.2:1.0


bjoern@raspberrypi:~ $ candump can0

hier bekomme ich leider gar keinen Output...

Ich hoffe jemand hat eine Idee, was ich machen kann.

Danke euch. Viele Grüße

platinum

**Reply by MyHomeMyData** [reply]
Das sieht recht seltsam aus. Hast Du den Adapter (mehrmals) ab- und angesteckt? Das könnte die mehrfachen Einträge im syslog erklären.
Die CAN-Schnittstelle empfängt überhaupt keine Messages. Entweder es gibt noch ein Hardware-Problem (Verkabelung, Adapter-HW), oder die Konfiguration des CAN-Bus stimmt nicht.

Würde mal folgendes probieren:
* CAN-Adapter abstecken
* Raspi neu starten
* CAN-Adapter wieder anstecken
* CAN-Bus initialisieren: `sudo ip link set can0 type can bitrate 250000 && sudo ifconfig can0 up`
* Schauen was `ifconfig -a` , `candump can0` und `lsusb` liefert

Fragen zur HW:
* Welchen CAN-Adapter hast Du?
* Wie sind Vitocal und VX3 verbunden und konfiguriert (als Master-Slave)?
* Wo ist der Adapter CAN-seitig angeschlossen?

**Reply by philippoo66** [reply]
@platinum83 mach auch noch einen `ip -details -statistics link show can0` damit man auch noch die Paket Statistik sieht

**Reply by platinum83** [reply]
> @platinum83 mach auch noch einen `ip -details -statistics link show can0` damit man auch noch die Paket Statistik sieht

Hallo ihr beiden,

nach dem Raspi Neustart liefert dmesg | grep CAN nun nur noch diese Einträge:

bjoern@raspberrypi:~ $ dmesg | grep CAN
[  107.014595] usb 1-1.3: Product: USB2CAN V3.3
[  107.142735] CAN device driver interface
[  491.207318] NET: Registered PF_CAN protocol family

Das scheint aber nicht das Problem gewesen zu sein.

ip -details -statistics link show can0 liefert folgendes: 

bjoern@raspberrypi:~ $ ip -details -statistics link show can0
4: can0: <NOARP,UP,LOWER_UP,ECHO> mtu 16 qdisc pfifo_fast state UP mode DEFAULT group default qlen 10
    link/can  promiscuity 0  allmulti 0 minmtu 0 maxmtu 0
    can state ERROR-PASSIVE restart-ms 0
          bitrate 250000 sample-point 0.875
          tq 250 prop-seg 6 phase-seg1 7 phase-seg2 2 sjw 1 brp 12
          gs_usb: tseg1 1..16 tseg2 1..8 sjw 1..4 brp 1..1024 brp_inc 1
          clock 48000000
          re-started bus-errors arbit-lost error-warn error-pass bus-off
          0          0          0          2          290538     0         numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 tso_max_size 65536 tso_max_segs 65535 gro_max_size 65536 parentbus usb parentdev 1-1.3:1.0
    RX:  bytes packets errors dropped  missed   mcast
            24       3      0       0       0       0
    TX:  bytes packets errors dropped carrier collsns
            24       3      0      10       0       0

Sieht das gut aus oder eher nicht? 
(Es scheint bei RX und TX ja zumindest etwas zu passieren.)


Morgen versuche ich nochmal ein neues Kabel und mache Fotos vom Anschluss.

Vielen Dank euch schon einmal.

Beste Grüße

platinum

**Reply by philippoo66** [reply]
moin Björn!

das sieht nicht gut aus: ERROR-PASSIVE heisst, das Ding hat dermassen viele Fehler 'gesehen', dass es nur noch passiv auf dem Bus lauscht (error-pass  290538 und dabei nur 3 Pakete empfangen...). 

Wenn die Baudrate stimmt (das tut sie ja anscheinend) hat das gewöhnlich Probleme bei der Bus-Physik als Ursache. Dir feht 'Abschluss'widerstand oder du hast zu viele, oder du hast ein Masse/Signal-offset Problem, oder du hast Hi und Lo irgendwo vertauscht oder so was in der Richtung... oft nur eine Kleinigkeit, die aber reicht, den ganzen Bus lahm zu legen ;-)

Grüsse!
Phil

**Comment by platinum83** [comment, marked as answer]
Hey Leute,

vielen Dank für eure Hilfe. Tatsächlich war es ein ganz blöder Fehler. Direkt nach der Klemme in meinem VX3 war die Leitung einfach abgebrochen. Das hatte ich vorher leider nicht gesehen.

Nochmal vielen Dank für eure Hilfe.

Eine kleine Frage hätte ich dann gleich noch. Bei manchen Datenpunkten wird eine Liste ausgegeben. Kann mach auch nur einzelne Teile der Liste ausgeben lassen?

z.B.

(.venv) bjoern@raspberrypi:~ $ open3e -r 0x6a1.1690 -v
0x6a1 1690 ElectricalEnergySystemPhotovoltaicStatus {"ActivePower cumulated": 5855.0, "RectivePower cumulated": 0.0, "ActivePower String C": 0.0, "RectivePower String C": 0.0, "ActivePower String B": 3032.0, "RectivePower String B": 0.0, "ActivePower String A": 2822.0, "RectivePower String A": 0.0, "OpMode": 2.0}

Ich möchte hier nur "ActivePower cumulated": 5855.0 in der Ausgabe haben. Geht das irgendwie?

Dankeschön, viele Grüße

platinum

**Reply by philippoo66** [reply]
schön dass du das Problem lokalisieren konntest! wie vermutet eine Kleinigkeit mit großer Wirkung ;-)

Das Ausgeben nur einzelner Sub-Elemente haben wir bisher nicht eingebaut. Wahrscheinlich werden wir das auch nicht machen, weil dann möchte der Nächste die Subs 3, 5 und 6 oder so. Das macht den Code unübersichtlich und die Chance für Falscheingaben gefolgt von dass der Code mit Fehlermeldung aussteigt steigt stark an...

Meist 'endet' die Geschichte ja in irgendeiner Hausautomatisierung, die dann ihrerseits irgendwelche spezifischen Objekte von Interesse im MQTT Tree 'herauspickt'.

Grüsse!
Phil

**Reply by MyHomeMyData** [reply]
@platinum83 Freut mich, dass es jetzt funktioniert.

In welcher Form bzw. mit welchem System verarbeitest Du die Daten denn weiter?

Je nachdem, ob Du die Option `-j` angibst oder nicht, werden die Daten im JSON-Format oder als Einzelelemente gesendet. Am besten Du schaust Dir beides mal in einem Tool wie  [MQTT-Explorer](https://mqtt-explorer.com/) an.

Falls Du ioBroker verwendest oder in Betracht ziehst: Dafür gibt es den Adapter [e3oncan](https://github.com/MyHomeMyData/ioBroker.e3oncan#readme), der die Daten automatisch aufbereitet. In ioBroker nach e3oncan suchen. open3e wird dabei nicht benötigt.

Gruß, Jürgen
